File: app/src/main/java/com/quigglesproductions/secureimageviewer/datasource/folder/IFolderDataSource.java
package com.quigglesproductions.secureimageviewer.datasource.folder;

import android.content.Context;

import com.quigglesproductions.secureimageviewer.SortType;
import com.quigglesproductions.secureimageviewer.appauth.RequestServiceNotConfiguredException;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.EnhancedFile;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile;

import java.io.FileNotFoundException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public interface IFolderDataSource {
    URL getFolderURL() throws MalformedURLException, RequestServiceNotConfiguredException;

    void getFilesFromDataSource(Context context, FolderDataSourceCallback callback, SortType sortType) throws MalformedURLException;

    void getThumbnailFromDataSource(Context context,FolderDataSourceCallback callback) throws MalformedURLException;

    default boolean moreItemsAvailable(){
        return false;
    }

    interface FolderDataSourceCallback{
        default void FolderFilesRetrieved(List<IDisplayFile> files, Exception exception){

        }

        default void FolderThumbnailRetrieved(Object thumbnailDataSource, Exception exception){

        }
    }
}

File: app/src/main/java/com/quigglesproductions/secureimageviewer/datasource/folder/IFolderDataSource.kt
package com.quigglesproductions.secureimageviewer.datasource.folder

import android.content.Context
import com.quigglesproductions.secureimageviewer.SortType
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile
import com.quigglesproductions.secureimageviewer.room.databases.unified.UnifiedFileDatabase
import java.net.MalformedURLException
import java.net.URL

interface IFolderDataSource {
    fun getFolderURL(): URL?

    @Throws(MalformedURLException::class)
    suspend fun getThumbnailFromDataSourceSuspend(
        context: Context,
        database: UnifiedFileDatabase
    ): Any?

    fun moreItemsAvailable(): Boolean {
        return false
    }

    interface FolderDataSourceCallback {
        fun FolderFilesRetrieved(files: List<IDisplayFile?>?, exception: Exception?) {}
        fun FolderThumbnailRetrieved(thumbnailDataSource: Any?, exception: Exception?) {}
    }

    enum class FolderSourceType {
        LOCAL,
        ONLINE
    }
}
